### ISSUE TITLE
paper #Q: Testing Self-Adaptive Software with Probabilistic Guarantees on Performance Metrics

### ISSUE THREAD

**Issue by kstolee** [researchart/fse20#112]
This issue is for discussion related to PR #16

Author contact: @ManCla

**Comment by kstolee** [comment]
## Authors

claudio.mandrioli@control.lth.se
github-id: ManCla (corresponding author)

maggio@cs.uni-saarland.de
github-id: martinamaggio

**Comment by art-reviewer** [comment]
[Paper](https://github.com/ManCla/fse20/blob/317526cf3dbf807f1dfd79749f52ff0880f30377/submissions/reusable/Mandrioli/paper.pdf)

**Comment by blind-reviewer-2** [comment]
Hi Authors (@ManCla),

Thank you for contributing to OpenScience by submitting your artefacts to be reviewed and made available to everyone. You have produced some incredible work both in your research as well as in the artefacts you have provided. Overall, I am very impressed by your artefact and it passes almost every criteria. You have applied for the "Reusable Badge", which in definition actually means that you have applied for all three badges. See the FSE2020 Artefact CFP [3] that reads for the badge of Reusable: "Available + Functional + very carefully documented..." This means, your artefacts must also be Available and Functional. Your artefacts do pass the criteria for both the individual requirements for Functional and Reusable, but as it stands, I don't believe your artefact is Available. For this, your artefacts must be "placed on a publically accessible archival repository." This includes services such as Zenodo [2]. This does not include GitHub (see comments below).

Please upload your artefacts to a publicly accessible archival repository, and link to that via the provided DOI in your paper and GitHub repo. After that, I believe your artefact passes the criteria for Available, Functional, and Reusable. Once this is done, I will change my Recommended Badges to include all three.

### Recommend Badges: None, Requires Available Artefacts

##### Available

Author-created artifacts relevant to this paper have been placed on a publically accessible archival repository. A DOI or link to this repository along with a unique identifier for the object is provided.

- I don't see a link to a repository where this data is stored. Please upload all relevant artefacts to an archive service such as Zenodo and link the provided DOI in your paper, as well as in GitHub.
- GitHub is not an archival service with the long-term mission to publicly provide it's data for the foreseeable future. As such, it is not a suitable replacement for services such as Zenodo [2].

##### Functional

*Documented*: At minimum, an inventory of artifacts is included, and sufficient description provided to enable the artifacts to be exercised.

- The README.md file in the SAVE folder is very well done. The description of the directories is a nice way to introduce the work.
- Typo in SAVE/README.md: "cxombinations"
- Typo in TAS/README.md: "funcitons"
- The README.md file in the TAS folder is similarly well done.

*Consistent*: The artifacts are relevant to the associated paper, and contribute in some inherent way to the generation of its main results.

- The artefacts provided output the main results of the paper, exactly. This is incredibly good work.

*Complete*: To the extent possible, all components relevant to the paper in question are included.

- The `videos` folder is too large to be included, but the authors have provided a link [1] that I tested and it does work. NOTE: I would prefer that all artefacts are publicly available in a repository designed to exist forever, such as Zenodo. However, I recognise that these files in particular are EXTRA files, and the scripts do execute with the given files.

*Exercisable*: Included scripts and/or software used to generate the results in the associated paper can be successfully executed, and included data can be accessed and appropriately manipulated.

- I recommend adding `sudo apt update` and `sudo apt-get update` to the beginning of your list of commands. This helps bring systems to a more updated place, often needed to run a list of apt-get commands. At the very least, it can help people get past a lot of issues that may arise. Additionally, running `sudo apt-get update` is sometimes necessary to update the packages after they have been loaded into the host machine.

- When running even `./unpack.sh`, there was a long waiting time. I would recommend adding a time estimation for each script that you know will take longer than just a few seconds. While I understand the difficulty in providing an accurate estimate, providing ANY estimate is a good start so that people know some sort of magnitude for waiting time. When creating these estimates, I would run all of your README and INSTALL steps on a standard machine designed just to test the replication process. Most people testing your replication packages will not have the case-specific hardware that you used.

  - For my MacBook Pro '15 running Docker with access to 6CPUs and 12GB RAM,  it took about 10 minutes. As a "standard" machine, I would say this is a good lower-bound on the time estimation required for this particular script with the subset of videos you have provided.
  - To increase the confidence of the reviewer/replicator, you can add a loading bar to long scripts, or, you can add instructions for how to check the progress of a script that is working file-by-file like yours. E.g.: "Open a new terminal window, navigate to the folder `SAVES/frames/` and look for the creation of sub-folders beginning with `unpacked_`. As the script runs, it will slowly create those folders and sub-artefacts."

- `./run.sh` has a loading bar, this is very nice feedback. Thank you.

  - I was able to successfully complete this command, in time. Due to obvious time constraints, I simply ran this for two random examples, to test the script. The first executed without issue, the second failed after ~150 minutes, and the third executed without issue. See detailed Installation and Execution steps below to see which two I chose.

  - Warning Encountered:

    ```bash
    /usr/lib/python2.7/dist-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
    ```

    It took 30 minutes for that to download, followed by the normal execution of the script

- Running `./figure.sh precomputed`

  - Error Encountered:

    ```bash
    root@4c6d4d0f43de:/home/SAVE# ./figure.sh precomputed
    This is pdfTeX, Version 3.14159265-2.6-1.40.16 (TeX Live 2015/Debian) (preloaded format=pdflatex)
     restricted \write18 enabled.
    entering extended mode
    (./Fig4.tex
    LaTeX2e <2016/02/01>
    Babel <3.9q> and hyphenation patterns for 3 language(s) loaded.
    
    ! LaTeX Error: File `standalone.cls' not found.
    
    Type X to quit or <RETURN> to proceed,
    or enter new name. (Default extension: cls)
    ```

    I got passed it by simply hitting "Enter"

  - Error Encountered:

    ```bash
    ! Package pgfkeys Error: Choice '1.14' unknown in choice key '/pgfplots/compat/
    anchors'. I am going to ignore this key.
    
    See the pgfkeys package documentation for explanation.
    Type  H <return>  for immediate help.
     ...                                              
                                                      
    l.13 \pgfplotsset{compat=1.14}
    ```

    I got passed it by simply hitting "Enter" for each error. Not sure where this error comes from, perhaps you can note the need for "Enter" to bypass any such errors.

- With MatLab installed, I was able to run `MAIN_tests.m`

  - I recommend that you add a few steps to the "In matlab run the script MAIN_tests.m in the /TAS/code directory" step. You have to select the folder in which Matlab should run, followed by loading the correct file, and then running the file. Perhaps "obvious" for someone who works with MatlaB a lot, but not obvious for the uninitiated.

- For the step: "To generate figure 3 from the paper you can now run the makefile" I recommend that you add the individual commands. Either add how to run the makefile (which didn't work on my machine for some reason), or add the individual commands, as there are only two of them

- You may wish to add some of my installation and execution steps to your readme. I only added commands when I needed to work through an error, so they may be essential to your scripts working on a fresh installation of Ubuntu.


##### Reusable

The artifacts associated with the paper are of a quality that significantly exceeds minimal functionality. That is, they have all the qualities of the Artifacts Evaluated – Functional level, but, in addition, they are very carefully documented and well-structured to the extent that reuse and repurposing is facilitated. In particular, norms and standards of the research community for artifacts of this type are strictly adhered to. 

- The artefacts presented are very well done.
- As stated earlier, the READMEs are very easy to follow and well done.
- The code is documented at both the function level and line level, where necessary.
- There are some places where comments are omitted, but I don't feel this leads to a lack of understanding of the code.
- The overall structure of the artefact folder structure and step-by-step instructions are incredibly well done and make it easy to follow.

[1] http://archive.control.lth.se/attic/research/videos.zip
[2] https://zenodo.org/
[3] https://2020.esec-fse.org/track/esecfse-2020-artifacts

### Full Installation and Execution Steps Taken

SAVE Algorithms

```bash
# Get the environment (OS) to run the code
# NOTE: Pulling "ubuntu" did not work in the following steps, but "ubuntu:xenial" did work. There is some issue with the latest ubuntu having access to all of the required packages.
docker pull ubuntu:xenial
docker run -it ubuntu:xenial

# Update system packages
apt update
apt-get update

# Install packages necessary for Artefact
apt-get install mplayer
apt-get install imagemagick
apt-get install python-imaging python-numpy python-scipy
apt-get install python-matplotlib python-cvxopt
apt-get install texlive-base texlive-pictures
apt-get install texlive-latex-extra

# Update all packages that were just installed
apt-get update

# Copy the program files into the running docker container
# Separate Terminal window, at the location of the SAVE folder
cd ~/GitRepos/AEC\ Artefacts/FSE20/Paper\ Q
docker ps  # To get the container ID for the next command
docker cp SAVE 3a7ff203fa9f:/home/

# Run the unpack script
cd /home/SAVE/
./unpack.sh  # Execution Time: 10 Minutes

# Checking on the running script, in a new terminal window
docker ps  # To get the container name for the next command
docker exec -it awesome_elion /bin/bash
cd /home/SAVE && ls -la  # Check if the "frames" directory has in fact been created
cd frames && ls -la  # To see if there are sub-directoris being created. This would signal progress

# Running the encoding script(s)
cd /home/SAVE/frames/ && ls -la && cd ..  # Check the existing folders for the next commands
./run.sh unpacked_480P random  # Execution Time: 13 Minutes
./run.sh unpacked_2160P integral  # Execution Time: Program Exited after 150 Minutes (Memory Issue?)
./run.sh unpacked_720P integral  # Execution Time: 26 Minutes
# NOTE: The first run of your ./run.sh script produced a warning and download that took 30 minutes to complete

# Running the figure script, with "precomputed"
./figure.sh precomputed

# Get the files out of the docker container so I can investigate the generated PDF
docker cp 3a7ff203fa9f:/home/SAVE/fig docker_export/
```

**Comment by gousiosg** [comment]
@blind-reviewer-2 Excellent review!
@allyourvirtualmachinesarebelongtous We would appreciate your review till Monday morning AoE.

**Comment by ManCla** [comment]
Thanks a lot for the time you spent reviewing our artifact and for the accurate feedback!

Since the discussion is supposed to start today I have created a new pull request where we address the different remarks. Also, I have made public the original artifact repo in my own github account  (with the associated Zeonodo repository).

Please let us know if there is more we should do.

**Comment by blind-reviewer-2** [comment]
Hi Authors (@ManCla),

Thank you for Zenodo DOI. I have updated my review above to include the DOI, and change my assessment to "Available & Reusable".

Reviewer

**Comment by kstolee** [comment]
Relevant PR #141 merged

**Comment by allyourvirtualmachinesarebelongtous** [comment]
Recommendation: Reusable

## Summary

The artifact includes necessary files, installation and execution instructions for the two case studies TAS and SAVE within the paper "Testing Self-Adaptive Software with Probabilistic Guarantees on Performance Metrics".

  - The artifact archive is available in a publicly accessible repository (Zenodo)

  - The archive (and GitHub repository) provides a detailed inventory and documentation of the required files, tunable parameters, 

   - It also provides pointers to a larger dataset of videos and how the specific results of the paper can be reproduced

## Available

- [x] The artifact is available on Zenodo, a publicly accessible archival repository and has a [DOI](https://zenodo.org/record/3896795).

## Functional

- [x] Documented: The artifact contains a detailed inventory of the data, installation instructions, and scripts required to reproduce results.
(There is a small typo in the `README.md`: Running the two case study -> case studies)

- [x] Consistent and Complete: The documentation in the artifact explicitly and clearly establishes the relationship between contents in the artifact to the results of the case studies in the paper.

- [x] Exercisable: I was able to exercise all the scripts on the provided data for the SAVE case study. I do not have a Matlab installation readily available to properly test the TAS case study. I tried executing it with Octave, but was not able to produce the results (I understand that these two are not trivially interchangeable, so I trust the results by @blind-reviewer-2). However, the process is well documented and makes the parameters to the tooling explicit. 

    - I created a basic Dockerfile in the SAVE directory (based on the documented instruction steps by @blind-reviewer-2 - thanks!). This may be helpful to support future reproducibility efforts:

```Dockerfile
FROM ubuntu:xenial

# Update system packages
RUN apt update -y
RUN apt-get update -y

# Install packages necessary for Artefact
RUN apt-get -y install mplayer && \
    apt-get -y install imagemagick && \
    apt-get -y install python-imaging python-numpy python-scipy && \
    apt-get -y install python-matplotlib python-cvxopt && \
    apt-get -y install texlive-base texlive-pictures && \
    apt-get -y install texlive-latex-extra

# Update all packages that were just installed
RUN apt-get update -y

# Copy the program files into the running docker container
COPY . /home/SAVE

# Run the unpack script
WORKDIR /home/SAVE/

RUN unpack.sh

CMD /bin/bash
```

## Reusable

Given the clear traceability from data, analysis, and results, as well as the very detailed instructions and documentation, I can additionally recommend the reusable badge.

**Comment by kstolee** [comment]
Great work, all! Looks like we have a fast consensus. 

@ManCla please cite the artifact in your paper by including the DOI in the references section.
